### PR TITLE
@ #10 | vagrant gatling auto restart when error

### PR DESCRIPTION
**Pull request by hieptranquoc**
connect #10
connect #11 
please check this PR! Thank you!

### CODE REVIEW

File: lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      def configure_node(settings, config)
+        synced_folders_settings = settings['vm']['synced_folders'] || []
+        return if synced_folders_settings.empty?
+
+        rsync_exists = false
+        synced_folders_settings.each do |synced_folder|
+          rsync_exists = true if synced_folder['type'] === 'rsync'
+        end
+
+        return unless TeracyDev::Util.true?(rsync_exists)
```
Comment (datphan) [inline lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb]: this look cleaner for me

```
rsync_exists = synced_folders_settings.find { |x| x['type'] == 'rsync' } # or x[:type]

return if rsync_exists.nil?
```

File: lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+        if gatling_rsync?
+          # To Ensure gatling-rsync don't run on start up
+          config.gatling.rsync_on_startup = false
+        else
+          return
+        end
```
Comment (datphan) [inline lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb]:
```
return if !gatling_rsync? # or unless gatling_rsync?

# To Ensure gatling-rsync don't run on start up
config.gatling.rsync_on_startup = false

....
```

File: lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        # TODO (hieptq) : should make 'vagrant rsync-auto' restart when crashed
```
Comment (datphan) [inline lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb]: should remove "TODO.."

File: lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+      def gatling_rsync?
+        gatling = false
+
+        if TeracyDev::Plugin.installed?('vagrant-gatling-rsync')
+          gatling = true
+        end
+
+        gatling
+      end
```
Comment (datphan) [inline lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb]:
```
def gatling_rsync_installed?
  return TeracyDev::Plugin.installed?('vagrant-gatling-rsync')
end
```

**Review by datphan — COMMENTED** [review]
@hieptranquoc pls check my comment

**Comment by hieptranquoc** [comment]
@datphan applied your comment! Thank you

**Comment by hieptranquoc** [comment]
Please check with step:
1. create `guest_test.sh` in workspace
```
#!/bin/bash
for i in {0..10000}
do
    touch $(pwd)/test.yaml
    echo "file created test.yaml"
    rm -rf $(pwd)/test.yaml
    echo "file deleted test.yaml"
done
```

2. create `host_test.sh` in workspace
```
#!/bin/bash
for i in {0..10000}
do
    touch $(pwd)/test.yaml
    echo "file created test.yaml"
    sleep 2
    rm -rf $(pwd)/test.yaml
    echo "file deleted test.yaml"
done
```
3. vagrant up
4. open 1 terminal tab, and run `vagrant ssh` , and then run `workspace/guest_test.sh`
5 open another terminal tab, and run `workspace/host_test.sh`

File: lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        # To Ensure gatling-rsync don't run on start up
+        config.gatling.rsync_on_startup = false
+
+        config.trigger.after :up, :reload, :resume do |trigger|
+          trigger.ruby do |env,machine|
+            begin
+              system("vagrant gatling-rsync-auto")
+              raise unless $?.exitstatus == 0
+            rescue
+              retry
```
Comment (datphan) [inline lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb]: should have a message to tell user about restart the rsync
```
puts "rsync crashed, retrying..."
```

**Review by datphan — COMMENTED** [review]
@hieptranquoc  pls check my comment
vagrant 2.1.2 does not working, but vagrant 2.2.2 does

**Comment by hieptranquoc** [comment]
@datphan applied your comment! Thank you. Please check again!

**Review by datphan — DISMISSED** [review]

File: lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
+require 'teracy-dev/config/configurator'
```
Comment (hoatle) [inline lib/teracy-dev-v05-compat/trigger/rsync_trigger.rb]: let's rename and move the `rsync_trigger.rb` to `lib/config/gatling-rsync-recovery.rb`
```ruby
GatlingRsyncRecovery < Configurator
```

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc pls check my comments

**Comment by hoatle** [comment]
@datphan's comments on this:

```
@hieptranquoc pls check my comment
vagrant 2.1.2 does not working, but vagrant 2.2.2 does
```

@hieptranquoc make sure the changes should work with vagrant >= 2.1

**Comment by hieptranquoc** [comment]
@hoatle I've updated! please check this PR again! Thank you!

File: lib/teracy-dev-v05-compat/config/gatling_rsync_recovery.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        # To Ensure gatling-rsync don't run on start up
+        config.gatling.rsync_on_startup = false
+
+        config.trigger.after :up, :reload, :resume do |trigger|
+          trigger.ruby do |env,machine|
+            begin
+              system("vagrant gatling-rsync-auto")
+              raise unless $?.exitstatus == 0
+            rescue
+              puts 'rsync crashed, retrying...'
```
Comment (hoatle) [inline lib/teracy-dev-v05-compat/config/gatling_rsync_recovery.rb]: use @logger.info here

**Review by hoatle — COMMENTED** [review]
@hieptranquoc LGTM, need to use logger.info instead of puts. need @phuonglm's approval.

**Comment by hieptranquoc** [comment]
I've applied your comment! Thank you!

**Comment by phuonglm** [comment]
```
==> iorad.local: Doing an initial rsync...
==> iorad.local: Rsyncing folder: /home/phuonglm/Projects/teracy/dev/workspace/ => /home/vagrant/workspace
==> iorad.local:   - Exclude: [".vagrant/", ".git", ".idea", ".#*", "node_modules", "bower_components", "iorad/node/static/downloadtutorials/html", "iorad/.tmp", "iorad/app/scripts/base/templates/dist", "iorad/app/scripts/editor/templates/dist", "iorad/app/scripts/editor/workers/dist", "iorad/app/scripts/player/templates/dist", "iorad/app/scripts/player/workers/dist", "iorad/node/static/_dist", "iorad/node/static/dist", "iorad/node/static/assets/css", "iorad/node/static/webstart", "iorad/node/static/data", "iorad/node/static/assets/avatars", "iorad/node/webpack-assets.json", "target", ".backup-info", "iorad/helm-charts", "iorad/scripts", "iorad/.cache"]
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

find /home/vagrant/workspace -path /home/vagrant/workspace/.git -prune -o -path /home/vagrant/workspace/.idea -prune -o -path /home/vagrant/workspace/.\#\* -prune -o -path /home/vagrant/workspace/node_modules -prune -o -path /home/vagrant/workspace/bower_components -prune -o -path /home/vagrant/workspace/iorad/node/static/downloadtutorials/html -prune -o -path /home/vagrant/workspace/iorad/.tmp -prune -o -path /home/vagrant/workspace/iorad/app/scripts/base/templates/dist -prune -o -path /home/vagrant/workspace/iorad/app/scripts/editor/templates/dist -prune -o -path /home/vagrant/workspace/iorad/app/scripts/editor/workers/dist -prune -o -path /home/vagrant/workspace/iorad/app/scripts/player/templates/dist -prune -o -path /home/vagrant/workspace/iorad/app/scripts/player/workers/dist -prune -o -path /home/vagrant/workspace/iorad/node/static/_dist -prune -o -path /home/vagrant/workspace/iorad/node/static/dist -prune -o -path /home/vagrant/workspace/iorad/node/static/assets/css -prune -o -path /home/vagrant/workspace/iorad/node/static/webstart -prune -o -path /home/vagrant/workspace/iorad/node/static/data -prune -o -path /home/vagrant/workspace/iorad/node/static/assets/avatars -prune -o -path /home/vagrant/workspace/iorad/node/webpack-assets.json -prune -o -path /home/vagrant/workspace/target -prune -o -path /home/vagrant/workspace/.backup-info -prune -o -path /home/vagrant/workspace/iorad/helm-charts -prune -o -path /home/vagrant/workspace/iorad/scripts -prune -o -path /home/vagrant/workspace/iorad/.cache -prune -o '!' -type l -a '(' ! -user vagrant -or ! -group vagrant ')' -exec chown vagrant:vagrant '{}' +

Stdout from the command:



Stderr from the command:

find: ‘/home/vagrant/workspace/iorad/.9577.yaml’: No such file or directory
chown: cannot access '/home/vagrant/workspace/iorad/.9577.yaml': No such file or directory

[TeracyDevEssential::Config::HostManager][WARN]: [["192.168.1.206 iorad.local", "192.168.1.206 iorad.local"], ["192.168.1.206 iorad.local", "192.168.1.206 iorad.local"]] are conflicted with each other, maybe you haven't clean the old VM setup, you can resolve this by go to the old setup repo then run: 'vagrant destroy' or 'vagrant halt && vagrant hostmanager' or you can open '/etc/hosts' and clean it manually.

==> iorad.local: Doing an initial rsync...
==> iorad.local: Rsyncing folder: /home/phuonglm/Projects/teracy/dev/workspace/ => /home/vagrant/workspace
==> iorad.local:   - Exclude: [".vagrant/", ".git", ".idea", ".#*", "node_modules", "bower_components", "iorad/node/static/downloadtutorials/html", "iorad/.tmp", "iorad/app/scripts/base/templates/dist", "iorad/app/scripts/editor/templates/dist", "iorad/app/scripts/editor/workers/dist", "iorad/app/scripts/player/templates/dist", "iorad/app/scripts/player/workers/dist", "iorad/node/static/_dist", "iorad/node/static/dist", "iorad/node/static/assets/css", "iorad/node/static/webstart", "iorad/node/static/data", "iorad/node/static/assets/avatars", "iorad/node/webpack-assets.json", "target", ".backup-info", "iorad/helm-charts", "iorad/scripts", "iorad/.cache"]
==> iorad.local: Watching: /home/phuonglm/Projects/teracy/dev/workspace
```
auto restart after crash, LGTM

**Review by phuonglm — APPROVED** [review]

**Review by hoatle — APPROVED** [review]

**Pull request merged**